### ISSUE TITLE
Confused as to what my sources.list should be on older Nexus 7

### ISSUE THREAD

**Issue by naturalstate** [offensive-security/kali-nethunter#1272]
### OS version (KitKat/Lollipop/Marshmallow/Nougat):
Marshmallow 6.0.1

### Built from repo (date and build command) or downloaded from website (links):
Downloaded Nexus 7 2013 zip from offensive nethunter download page

### Output of `cat /proc/version` in adb shell or NetHunter Terminal:
Linux version 3.4.0-gcc51ee3-dirty (root@kali) (gcc version 4.7 (GCC)) #4 SMP PREEMPT Sat Dec 12 00:29:55 UTC 2015

## Issue:
I'm having trouble updating with the sources.list file that comes with nethunter. I've tried my best to search around and see that other folks have similar issues but I still can't get it to work.

My sources.list has "sana" which I understand isn't supported anymore. So I modified the sources.list like the official sources are listed on kali's website.  But I'm a bit confused since it's a nethunter if it is supposed to be like that or different. I saw some people say that nethunter shouldn't use the regular Kali sources as it will break the nethunter install. I'm so confused.

I also removed the nethunter app that came with the iso and installed the 2019.1. 
When I try to add metapackages with this app it now says "The following packages have unmet dependencies"

1. What should my sources.list file look like with the older phone that I am using?  
2. If I get the sources file fixed, will the app intall metapackages successfully? 
3. Or can If I want to install the Kali top 10, can I use the command : apt update && apt install kali-linux-top10  ? 

I've installed nethunter over 10 times now. Can someone help please? 
Thank you!

**Comment by 0E800** [comment]
You dont need to do all that. You just need to manually install the Kali Archive Keyring

`wget https://http.kali.org/kali/pool/main/k/kali-archive-keyring/kali-archive-keyring_2018.1_all.deb`

`dpkg -i kali-archive-keyring2018.1all.deb`

`apt update && apt upgrade`


BTW - I am making a updated ROM and Kernel for N7 2013. Possibly I can use you as a tester?

**Comment by naturalstate** [comment]
Thanks for the help.
I tired running these commands on my existing install, which may already be messed up.
The first one worked fine. But the second one (dpkg) has error. Is that last part correct? the .1all.deb ? I notice that is different than the first command.

I'm going to reinstall nethunter. And try these commands first. And based on your explanation, I won't modify the sources.list file. 

Yes, I would be happy to test your ROM.

**Comment by naturalstate** [comment]
So I just reinstalled nethunter.
I ran the wget successfully. And then the dpkg successfully.
But when I run apt update && apt upgrade I get 404 errors.
If I go to the nethunter app and try to add metapackages I get a "failed to fetch..." It does at first, list a bunch of packages, so I thought it was going to work at first, but then fails. 

So I have a problem with my sources file, yes?

My source file is:
deb http://http.kali.org/kali sana main non-free contrib
#deb-src http://http.kali.org/kali sana main non-free contrib
deb-src http://security.kali.org/ sana/updates main contrib non-free
#deb-src http://security.kali.org/ sana/updates main contrib non-free

Sana doesn't work anymore right? So what do I need to do to fix my sources file?

Thank you so much for help, really butting my head against a wall here.

**Comment by 0E800** [comment]
Doh! Sorry about that. In the future, try using the [Tab] key to autocomplete the file names.
That my bad though, you are correct, dpkg would not be able to install a package if the filename is typed incorrectly.

dpkg kali-[Tab] would have autocomplete the filename and saved you a few finger jabs.

I corrected the instruction.

Theres another command I am missing. It corrects the sources. Ill get you the source or the command.

One moment.

**Comment by 0E800** [comment]
While I locate info, may you try `apt dist-upgrade `

**Comment by naturalstate** [comment]
After changing my sources to the official kali sources listed on their site, running apt update && apt upgrade appears to be working! And I see it's referencing armhf. Not sure I saw that before, so that seems like the right track.

But right before I tried that, the nethunter app to download metapackages had 404, it wouldn't even list the packages that it did earlier before I change the sources.list. 
So it seems only one or the other works (nethunter app metapackages download or apt get / upgrade)

It's on 18% now, and looks like it's going to take a while.
I wonder if the nethunter app will work after this.

**Comment by naturalstate** [comment]
If the nethunter app to install metapackages doesn't work, I'm wondering if I can use something like this to install specific metapackages? I'm just a little unclear if that's going to install them like the nethunter app would. 
apt update && apt install kali-linux-sdr

**Comment by 0E800** [comment]
Your using and old build. You should replace the kalifs-full.xz with a  more recent build.
You will need to also download newer builds of term-nh.apk and nethunter.apk

Check the issues, there are similar users with problems.. the apk files are pretty much universal, you just want the more current ones.

I am banging my head against the keyboard trying to build a Pie kernel for the Pixel XL and I am unable to get everything within the size of the boot partition and I am frustrated so I cant really do the leg work, but google and specific but few words you will find your way.

I need a beer.

**Comment by 0E800** [comment]
Using the chroot nethunter app to install the meta packages is the same as apt install kali-wireless-meta [or however its named] from within Kali chroot environment that you access from term-nh. 
True only when your nethunter and terminal are updated to match the kalifs

**Comment by naturalstate** [comment]
Ok. You've given me a lot to research. Thanks

I didn't know that I could have Oreo and run nethunter on the Nexus 7. I mean, on the nethunter download page the only option available for Nexus 7 is Marshmallow & lolipop. So I just assumed that's what I would need.

My updates are still going. So I'm going to wait to see what happens first.

Really appreciate the help!

**Comment by 0E800** [comment]
I broke out my N7, I got a few 2012 and a 2013 model. I really want to get the 2012 updated but its limited and the majority of my developemnt device are cracked screened and some even require mouse and keyboard hook ups in order to navigate.

Ultimately my personal aim is to jailbreak peoples dependency of requiring a display on a smart device.
Its like flying cars. Its possible, but we still stuck on 4 wheels.
Having no display on our smart devices forces us to interact with the device in a way that doesn't alter human evolution into having slouched neck-bent spinal cords.
Forces manufactures to realize we are no longer toddlers amused by buttons and the satisfaction of pressing them.
Provides forward development into AI and maybe Alexa and Hey Google will grow a pair since most of our interaction with the device/phone will be auditory, verbal, gesture, possibly the display becomes a pressure sensitive pad that responds to pressure / time to operate.

Apologies, the 40oz got me sided tracked with my crusade to return honor to the discarded devices that are not worthless.

tl:dr,

I am going to test my builds on my N7-213 wifi and upload a working build that hopefully will be well received.

**Comment by naturalstate** [comment]
Well I can test something if you want. 

Enjoy the 40! haha

So the updates finally finished. Although they cancelled at one point with:
E: Sub-process /usr/bin/dpkg returned an error code (1)
I ran apt-get update && apt-get upgrade, and they finished after that.

But when I tried to install metapackages via nethunter app, it found a bunch of packages, but then has a ton of "failed to fetch".

So what should I do? Upgrade to oreo? But then, what nethunter iso do I use? Because there aren't any listed for N7 2013 for oreo?

**Comment by 0E800** [comment]
Don't erase yet. I. Uploading newer apps

**Comment by 0E800** [comment]
https://drive.google.com/file/d/17pk5QX-liyOrDQOsLTd7MsfKuKSckKiS/view?usp=drivesdk

Apps.zip from magisk nethunter mod 3.2019

**Comment by naturalstate** [comment]
Thanks, but the link isn't working. It says it violates google's TOS.

**Comment by 0E800** [comment]
Okay I'll try again and put a password on it .

**Comment by 0E800** [comment]
https://drive.google.com/file/d/1u6Zye2CpVeUlf6OcqnORyIdNXykr6NBb/view?usp=drivesdk

The password is KALI [except is all lowercase and the the i is a one

**Comment by naturalstate** [comment]
Thanks. I installed everything following your instructions. But I'm still having issues installing metapackages.
The termnh crashed the first time that I launched it, similar to what I experienced when I first installed. But the second time it worked. But when I go to Nethunter app and try to install a metapackage like wifi tools, it says what packages it will install, and then says what packages will be installed, removed and upgraded and then there a bunch of "failed to to fetch" and temporary failure of kali.org.
I then tried to ping google.com in that same terminal session and it also fails. Weird. That's gotta be part of the problem right?

If I open another terminal, I can successfully ping google and run apt-get update / upgrade.

Ideas? Should I just start over and try to get Oreo and a later nethunter version?

**Comment by naturalstate** [comment]
I also just tried to install a metapackage using apt update && apt install kali-linux-wireless and it fails with "failure to resolve".

**Comment by 0E800** [comment]
You have more than one source.lst possibly .

Check sources .LST and also check sources.list.d/ any files in there to
make sure. Are you completely wiping your device between builds and
restoring to stock before flashing again?

You may have more than one chroot?

**Comment by 0E800** [comment]
Gimme a moment I'm getting my kid to bed and then I gotta search through my
USB collection for my n7 files.

I'll do all the legwork and then upload the build and kernel.

Good job sticking with it. Its not insanity because you are making lateral
progress even if it seems like quick sand

**Comment by naturalstate** [comment]
No, maybe I'm not.
I'm just rebooting to TWRP recovery and reinstalling the nethunter image. That might make sense that I could have 2 source files.
I'm really new to android roms and flashing and all of that. I'll research how to restore to stock. I'm assuming that I should use the nexus root toolkit to do this?

**Comment by naturalstate** [comment]
I did a bunch of these errors after installing the nethunter image in TWRP. 
error opening: '/data/local/nhsystem/kali-armhf/dev/ram12 (not a directory)

Really appreciate your help on this

**Comment by 0E800** [comment]
Yeah you need to use the wipe all data option , it's separate from the stock factory reset wipe.

Complely wipe and don't reboot .
First you need to go to advanced and mount the system folder.
Then after you full data wipe , go back a few times and click reboot to recovery and do a full data wipe again. The errors should be gone

**Comment by 0E800** [comment]
Use NRT to flash back to stock 5.1.1 I think .

Then you can flash new ROM.

I'm getting the zips organised but feel free to flash a ROM from XDA if u want the practice . you will see in the rom install instructions that most ROMs follow the same instructions

**Comment by 0E800** [comment]
Install instructions.

Full data wipe and return to stock
Boot to TWRP recovery 
Install the ROM
Reboot back to recovery 
Flash sysrepart.zip
Do not reboot 
Go back and install Gapps pico
Reboot and go through initial android setup
Reboot to recovery 
Install custom Kali bundle

Please be patient I have about 6 different Oreo and Nougat builds and its been a year since I last touched em so you may have to mix and match to find what works best.

This is just the beginning. Wait till your up and figuring out the nuances of getting the external WiFi to work , the cords that charge but dont do data, the weak microusb connector that is temperamental and the reboots whenever the WiFi gets unplugged .

I'll try to get you up to speed. I'm using this as a distraction from the Pixel phone that is being a brat.

**Comment by 0E800** [comment]
@naturalstate 

Here you go man, rough draft but you should be able to muster through it.

https://github.com/0E800/Nexus-7-2013-Flo-Experimental/releases/tag/Flo-AIO-KaliPWN

One big ol' file that handles everything.
If its not stable enough for you, I can try upload the Nougat version.

**Comment by naturalstate** [comment]
Thanks so much, this looks perfect!
I'm working on it now.

**Comment by naturalstate** [comment]
So after rebooting to recovery in TWRP, I wiped by going to format and typing in "yes". I  did this 3 times and rebooted back to recovery each time. 
I realized that I need to boot into OS to turn MTP on and USB debugging so I can copy the KaliPWN_OREO.zip to the device to flash it. And I see that I don't have root anymore. I was wondering if I was going to be unrooted.
So I need to root with Nexus Root Toolkit again right? 
After that do I need to install busybox again?

**Comment by naturalstate** [comment]
Well shit, tried to root device again and went through all the steps that I did the first time. But now I'm getting message in Nexus Root Toolkit that my ADB device is found but it's unauthorized. It says to renable MTP & USB debugging, restarting device, and plugging into a different USB port, but I'm not getting the prompt to accept the RSA key. It won't let me flash to stock without this being taken care of either. Any ideas?

**Comment by 0E800** [comment]
Try this.

Boot into TWRP
Mount System
Goto Wipe -> type Yes wipe.
If not errors then reboot back to recovery
Do the normal factory swipe wipe
Put in USB drive with KaliPWN on it.
Mount it so you can see files on USB
Goto Advanced -> File Manager
Copy KaliPWN to the device directly.
Then take out usb.
Install / Flash KaliPWN

Might not need to mess with NRT

**Comment by 0E800** [comment]
The build has root already integrated
It does not install a icon or prompt you if you want an app to use it. It just auto-roots.

You can test this by opening up FX File Manager and going to System(root) 
If you dont have root, it will error on you and you wont see any files.

**Comment by naturalstate** [comment]
Ah, ok. Cool.
When I go to mount in TWRP and select System. Should I also have Data & Cache selected too?

**Comment by 0E800** [comment]
Yes, you want all of em mounted so that the info gets wiped. Im not sure if the wipe works if they are not mounted.

**Comment by 0E800** [comment]
ill be back in like 30 minutes. Try to follow the instructions on the Release page.
I was passing out but really did type it up as I was doing it myself on my own N7.

**Comment by naturalstate** [comment]
I'm on the roll a d00be stage and updating. So far so good!

**Comment by 0E800** [comment]
Excellent. I am really hoping that someone else can it to work. Then the 5+ years I have been baby stepping my way to being able to create this build will not goto waste. 👍

**Comment by naturalstate** [comment]
Finished the apt update && apt autoremove with no problems. And rebooted to be safe. Now trying to install wireless tools metapackage and so far so good. It's downloading everything now. Never got to this point!

**Comment by 0E800** [comment]
Awesome man. Wait on installing the GAPPS part.

I need to test a more current GAPPS for Oreo build. The PlayStore keeps crashing.

Also we will probably need to update the Nethunter and Term-NH APK packages as those are a year old and I see new builds have been compiled.

**Comment by naturalstate** [comment]
I do appreciate the police scanner, I saw that. Haha.
GAPPS as in google apps like drive, gmail etc? I'm fine with not having that. 
Which nethunter & term-NH apks just the newest ones? Those are pretty universal and should work on any system right?

Anything else that I should be aware of? 
It should be fine for me to continue updating with apt-get update && apt-get upgrade?

Any other apps that you've included that I should be aware of or may find useful?

**Comment by 0E800** [comment]
Yea. My bad I missed some of your posts.

#Big FYI#

If you turn on USB debuging you will lose root and HID Keyboard.   Not sure why but the silver lining is that your device is more secure without adb.

**Comment by 0E800** [comment]
I just discovered that someone has a buid of Pie for N7 2013. I'm gonna test and see if I can use it.

**Comment by naturalstate** [comment]
And so does that mean no google play either? So if I need to install an app just get the apk I guess? 

After I successfully installed the wifi metapackage, I don't see any of the tools. Weren't they supposed to create a new folder with the new tools? Since I've never got it to work, I'm not quite clear.

**Comment by 0E800** [comment]
Try installing the nethunter and apps I password protected they should be more recent. Compare the versions .

**Comment by naturalstate** [comment]
K, so unistall the current nethunter app and then install the one that you sent zipped up last night?
They metapackages are supposed to create new folders with tools though right? Can't find a lot of info about what they are actually supposed to do or if there are supposed to be new tools accessible in nethunter app.

**Comment by 0E800** [comment]
All the apps I think are useful. 2nd line will give you a free SMS phone number in just about any stare.
Lucky patcher will help you remove adds and give you paid for access to some apps .
Macrodroid allows you to set up rules that automate functions . like everytime my phone notices a Linksys WiFi connection it turns off WiFi and turns on Bluetooth and sends a email.

**Comment by naturalstate** [comment]
Yes, I think I googled every one to become familiar. Looks like a nice collection!

**Comment by 0E800** [comment]
No I think you misunderstand .

Nethunter has some android native tools but mostly its a bunch of helping wrapper scripts for managing the Kali chroot environment .

The metapackages are mostly command line tools

**Comment by naturalstate** [comment]
Ok cool, I did misunderstand. Thanks for clearing that up. I thought after you installed a metapackage it was going to create new icons for the new tools it installed. 

Is there any reason to upgrade the Nethunter app? I have a 2019.1 apk. Kind of scared to mess with it at this point, now that I've finally gotten mostly everything to work. 

Dude, you are the man! I absolutely would have not have gotten this thing to work without you! Is there anything that you want me to test for you? Or anything I can do in return? Would buy you a beer if you were in Colorado!

**Comment by 0E800** [comment]
You should ask yourself what you want the device to do for you. Then you learn the tools that help you achieve the goal. 
Possibly its more efficient for you to perform the tasks from a tablet. You wouldnt go scuba diving before you knew how to swim.. Right ?

**Comment by 0E800** [comment]
Its payment enough that you didn't wuss out on me. You deserve the pat on the back for sticking with it.

I think you will get a kick out of the USB keyboard .

Just google the names of the tools and follow some tutorials

**Comment by naturalstate** [comment]
Haha. Thanks a million man!
One last question, what tool do you recommend to do a full backup of everything so that If I screw something up I can just restore to what I have now? TWRP?

**Comment by 0E800** [comment]
This what I'm gonna make next. 

https://whitedome.com.au/re4son/sticky-fingers-kali-pi/

**Comment by 0E800** [comment]
I'm the wrong person to ask man.

I don't trust backups and always felt like i was gonna mess it up trying to back it up.

**Comment by 0E800** [comment]
Check this out. 

https://www.amazon.com/gp/aw/d/B010O1G1ES/ref=sspa_mw_detail_0?ie=UTF8&psc=1

Little hack buddy

**Comment by 0E800** [comment]
Forgot , you should enable developer tools and enable the advanced reboot options.

Also in settings you have a lot of customizations available from AICP Extras.

**Comment by naturalstate** [comment]
Cool. I just recently got into some hardware tinkering. Might check that out. 
Thanks for the tip and all the help!

**Comment by naturalstate** [comment]
If you're curious, one thing that I've encountered is trying to run an nmap scan started either in nethunter app or kali CLI crashes and reboots device.

**Comment by 0E800** [comment]
There is a known issue of the device rebooting when network devices like external usb wifi is disconnected from the device.

For example. I plug in my atheros usb injection supported external wifi into the Nexus. 
I unplug it... I plug it back in.. no effect on the N7
I plug it in, I run `airmon-ng start wlan1`
If I unplug the external wifi, or dont take into account how lose the OTG cable fits the N7, as soon as the device thinks the external wifi is not longer physically plugged in, the device is going to reboot.

The issue threads have multiple issues like this. There are some work arounds, like using alternative commands that accomplish what airmon-ng does that wont crash the system.

Can you record the steps I can take to replicate the crash? See if I can isolate the reason?

**Comment by emprise1976** [comment]
I have also tried the build and can confirm the reboot of the device after using nmap or zanti.

Thank you for your engament.

**Comment by naturalstate** [comment]
I launch a kali shell and run "nmap -v 192.168.1.131" and it immediately locks up and then a few seconds later reboots. I don't have any wifi cards plugged in. My device is connected over wifi. And my device can successfully ping that device. 
Really hope I can get nmap working, that's a pretty basic feature that I need to work.

**Comment by 0E800** [comment]
I agree. Wireshark being of equal importance.

I will see what I can do. Thank you for continuing to help QA.

**Comment by 0E800** [comment]
@naturalstate 

If you don't mind, may you review:
 https://security.stackexchange.com/questions/69180/nmap-crashing-embedded-system-in-strange-way

Please attempt a few of the examples to verify if any nmap command causes the reboot or if its certain nmap options being used.

Its a heavy ticket Friday for me so I have to appease to the mob of clients before I can do what I want, I appreciate your patience .

**Comment by naturalstate** [comment]
I ran tshark, and it seemed to work.
Ok, I'll check it out. Thanks

**Comment by 0E800** [comment]
@naturalstate 

If you are using `nmap -v`  this may also be the reason why it crashes:

https://github.com/nmap/nmap/issues/1112

**Comment by naturalstate** [comment]
It seems that any nmap command crashes. I've ran just a nmap 192.168.1.131 and it crashes. I made a post on the stackexchange post.

**Comment by 0E800** [comment]
Thank you. Ill try a few things.

Also am testing the various Google Play and Services APKs so you can use the store

**Comment by 0E800** [comment]
Verified Nmap reboot issue.
I used the Nethunter option Nmap Scan.

I think I also have a proper Google Services APK if you want to be able to use the Store and other Google services. 

I am attempting to build a Android 9 (Pie) build for N7-2013 to see if it can be Nethunter enabled.

**Comment by 0E800** [comment]
@naturalstate 

Check this out:
https://forums.kali.org/showthread.php?43300-Nethunter-on-OnePlus-One-several-odd-things-I-cannot-find-answers-to-on-forum-or-in-Google

I replied to the tread so hopefully it will get resolved. At-least its not unique to my build.

**Comment by 0E800** [comment]
If I run `nmap -sL 192.168.1.1/24`  it doesnt reboot.

Will try to reduce the options to see what is causing it.

**Comment by naturalstate** [comment]
-sL also did not crash. I started trying others, went to -sn and it crashed. -sL isn't really doing any scanning, it's only listing ports right?

**Comment by naturalstate** [comment]
Ran -sL and it also didn't reboot. Started going down the list and ran -sn and it crashed. -sL only lists targets right? It's not doing any scanning.

**Comment by 0E800** [comment]
I think its the rate or speed it's trying to scan . but i don't know exactly what ~v does. It probably a combo of commands

**Comment by naturalstate** [comment]
-v is for verbose mode, it's suppose to show you what nmap is doing. If you do -vv it shows more. And finally you can do -vvv for the most verbose.

**Comment by 0E800** [comment]
Yeah, sorry what I meant to say was the default scan settings when no options are used.

Like at what rate does is sent out pings etc.

Good news is I was able to build Android Pie for the Nexus 2013, now I just need to patch the kernel to support HID Keyboard and wifi injection. I was impressed that the OS seemed pretty responsive.
 
As for nmap, I am fairly certain that https://github.com/nmap/nmap/issues/1112 is the answer.
I havent been able to verify if the nmap version on kalifs includes that patch.
I will try to compile nmap with that patch and test, if it resolves the issue then ill include that version.

**Comment by emprise1976** [comment]
I suspect that this is a general problem with the kernel and / or the Wifi drivers. The APP zAnti is also affected. 
https://www.zimperium.com/zanti-mobile-penetration-testing

**Comment by 0E800** [comment]
If anyone wants to  beta test:

Android Pie Nethunter for Nexus 2013 Wifi (Flo) (have not tested Deb)

https://github.com/0E800/Nexus-7-2013-Flo-Experimental/releases/tag/RR_PIE_Flo_Kali

**Comment by emprise1976** [comment]
nmap is also not working with Android Pie Nethunter for Nexus 2013
Installation was without any problems

**Comment by 0E800** [comment]
@emprise1976

Yeah I noticed that. Haven't tried building nmap yet.  

Apologies there is a issue with changing wallpapers. I believe its a permission issue and am testing .I'll upload a new build once I have identified solution .

**Comment by emprise1976** [comment]
Also zAnti is not working

**Comment by 0E800** [comment]
Does zanti use nmap or libraries ? The website seems to be a paid for
service

**Comment by emprise1976** [comment]
I assume that nmap is used here, unfortunately it is not visible in the APP, but files are loaded the first time they are executed.

**Comment by emprise1976** [comment]
zAnti uses namp 7.31

**Comment by 0E800** [comment]
Its been a while. Theres as way to install an older version of nmap and not
interfere with apt. Else you end up messing with Kali meta depends.
Would like to verify if earlier versions work.
Maybe I'll install an older kalifs just to test.

**Comment by emprise1976** [comment]
@0E800 Any News on the NMAP Problem? Is there anything I can help with?

**Comment by 0E800** [comment]
The new build (r19) doesnt crash the device, but still gives a Segmentation fault error.

It would be helpful if you could try testing a separate nmap build.

I found: https://github.com/kost/nmap-android/releases/download/v7.31/nmap-7.31-binaries-armeabi.zip
Opened the archive and replaced nmap in:   
 `/System(Root)/data/local/nhsystem/kali-armhf/bin/`

But when I try to run nmap, it give me a permssion denied as though root wasnt good enough.

So yeah, if you could find another build, shoot, even try nmap from an older release just too rule out if its the application or kernel.

**Comment by naturalstate** [comment]
Any updates on this?

**Comment by 0E800** [comment]
@naturalstate @emprise1976 

I was able to get nmap working using an older nmap for android app call anmap.apk

https://code.google.com/archive/p/anmap/downloads

I will keep testing.
